### ISSUE TITLE
decode_csv is extremely slow

### ISSUE THREAD

**Issue by sgirkin** [tensorflow/tensorflow#11713]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
I tested on both MacOS 10, Centos 7, https://pastebin.com/LsADQ89s

- **TensorFlow installed from (source or binary)**:
Tensorflow 1.2.1, also tried with tensorflow compiled from sources

- **TensorFlow version (use command below)**:
('v1.2.0-5-g435cdfc', '1.2.1'), also compiled and tested from master

- **Python version**: 
2.7.13

- **CUDA/cuDNN version**:
no

- **GPU model and memory**:
no

- **Exact command to reproduce**:
see above

### The problem
I noticed that a huge amount of training time CPU is idle. I isolated the problem to the code snippet bellow, where I create a string tensor using tf.constant with N rows each of them contains comma separated list of randomly generated numberical values, then I decode these string tensor to rank2 tensor of floats using 3 different approaches. 

I discovered following:
decoding tfrecords with *tf.parse_example* is quite fast
decoding csv with *decode_csv* is ~20 (4 cores) 60 (16 cores) times slower than parse_example and it utilises only 1 core on my machine
decoding csv with *tf.string_split* + *tf.string_to_number()* is ~20-60 times slower than parse_example. 
It seems *tf.string_to_number()* makes it slow, if I comment out the line with *tf.string_to_number* it is 2-3 times slower than parse example

The code I used for that:
https://pastebin.com/ZR152y5M

### Source code / logs
Running parse_example
Total time: 2.01925897598

Running string_split
Total time: 5.52524709702

Running decode_csv
Total time: 22.6595089436

**Comment by reedwm** [comment]
It makes sense for `tf.decode_csv` to be slower than `tf.parse_example`, since `tf.train.Example`s are designed to be decoded fast while CSVs are not. When training, I would recommend first converting your data into `tf.train.Example`s.

If you find that TensorFlow decodes CSVs significantly slower than some other CSV decoder implementation, then reopen this issue.

**Comment by sgirkin** [comment]
Hi,

I implemented decoding csv using _tf.py_func_ and numpy (test_numpy_csv function in the code below). It seems even reading with numpy is faster than decode_csv. 

https://pastebin.com/rTZjCwGH

So far parsing csv using py_func+numpy is ~5 times faster than using decode_csv on this particular synthetic data (1000 examples with 100 float columns)

**Comment by reedwm** [comment]
Thank you for performing that benchmark. Since `tf.decode_csv` is significantly slower than numpy's `np.fromstring`, I'm reopening and marking as contributions welcome.

**Comment by AKindyakov** [comment]
I would like to reserve it if I can.

**Comment by AKindyakov** [comment]
If what Valgrind has said me is true, the biggest problem is a slow string to float conversion in [core/lib/strings/numbers.cc](https://github.com/tensorflow/tensorflow/blame/master/tensorflow/core/lib/strings/numbers.cc). 
It seems to be true. There are two string copying, unordered_map using and so on. I can adjust the [double-conversion](https://github.com/google/double-conversion) library here. It essential faster.

There are some inconsistencies with TF str-to-double functions. Double-conversion library has case sensitive way to specify special values ```NaN``` and ```INF```. I can create a simple patch to fix case sensibility. Another problem is a length of ```infinity``` word. In TF you can write ```inf``` and ```infinity```, but double-conversion supports only one. Numpy support only one too (```inf```). And may be TF can do the same.

What do you think about it? Can I use this lib here?

**Comment by reedwm** [comment]
@cwhipkey any thoughts on using the double-conversion library? The current function used, locale_independent_strtonum, is not intended to be fast.

**Comment by AKindyakov** [comment]
I've made some changes here for the sake of experiments. With the double-conversion library, decode_csv works almost the same time as numpy csv parsing. I'll send PR, please look at it.

**Comment by Androbin** [comment]
Has the PR sent sufficiently improved performance?

**Comment by yongtang** [comment]
@Androbin I looked the git history. The PR #12102 by @akindyakov (👍 ) was merged, and later reverted in PR #15133.

Later the change was re-introduced in PR #18746 and finally merged. For that I think the issue is resolved.

During the above mentioned process, the version used by cmake and bazel became out of sync. I created a PR #19726 to address the out of sync issue.

Other than that I think this issue could be closed.
